runners/apex/src/main/java/org/apache/beam/runners/apex/ApexYarnLauncher.java

```diff
@@ -0,0 +1,379 @@
+import java.util.jar.JarFile;
+
+import org.apache.apex.api.EmbeddedAppLauncher;
+import org.apache.apex.api.Launcher;
+import org.apache.apex.api.Launcher.AppHandle;
+import org.apache.apex.api.Launcher.LaunchMode;
+import org.apache.apex.api.Launcher.LauncherException;
+import org.apache.apex.api.Launcher.ShutdownMode;
+import org.apache.apex.api.YarnAppLauncher;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.SerializationUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Proxy to launch the YARN application through the hadoop script to run in the
+ * pre-configured environment (class path, configuration, native libraries etc.).
+ * <p>
+ * The proxy takes the DAG and communicates with the Hadoop services to launch
+ * it on the cluster.
+ */
+public class ApexYarnLauncher {
+  private static final Logger LOG = LoggerFactory.getLogger(ApexYarnLauncher.class);
+
+  public AppHandle launchApp(StreamingApplication app) throws IOException {
+
+    List<File> jarsToShip = getYarnDeployDependencies();
+    StringBuilder classpath = new StringBuilder();
+    for (File path : jarsToShip) {
+      if (path.isDirectory()) {
+        File tmpJar = File.createTempFile("beam-runners-apex-", ".jar");
+        createJar(path, tmpJar);
+        tmpJar.deleteOnExit();
+        path = tmpJar;
+      }
+      if (classpath.length() != 0) {
+        classpath.append(':');
+      }
+      classpath.append(path.getAbsolutePath());
+    }
+
+    EmbeddedAppLauncher<?> embeddedLauncher = Launcher.getLauncher(LaunchMode.EMBEDDED);
+    DAG dag = embeddedLauncher.getDAG();
+    app.populateDAG(dag, new Configuration(false));
+
+    Attribute.AttributeMap launchAttributes = new Attribute.AttributeMap.DefaultAttributeMap();
+    launchAttributes.put(YarnAppLauncher.LIB_JARS, classpath.toString().replace(':', ','));
+    LaunchParams lp = new LaunchParams(dag, launchAttributes);
+    lp.cmd = "hadoop " + ApexYarnLauncher.class.getName();
+    HashMap<String, String> env = new HashMap<>();
+    env.put("HADOOP_USER_CLASSPATH_FIRST", "1");
+    env.put("HADOOP_CLASSPATH", classpath.toString());
+    lp.env = env;
+    return launchApp(lp);
+  }
+
+  protected AppHandle launchApp(LaunchParams params) throws IOException {
+    File tmpFile = File.createTempFile("beam-runner-apex", "params");
+    tmpFile.deleteOnExit();
+    FileOutputStream fos = new FileOutputStream(tmpFile);
+    SerializationUtils.serialize(params, fos);
+    fos.close();
+    if (params.getCmd() == null) {
+      ApexYarnLauncher.main(new String[] {tmpFile.getAbsolutePath()});
+    } else {
+      String cmd = params.getCmd() + " " + tmpFile.getAbsolutePath();
+      ByteArrayOutputStream consoleOutput = new ByteArrayOutputStream();
+      LOG.info("Executing: {} with {}", cmd, params.getEnv());
+
+      ProcessBuilder pb = new ProcessBuilder("bash", "-c", cmd);
```

Comment:
I think `params.getCmd()` and `tmpFile.getAbsolutePath()` should be additional varargs here, not pre-concatenated.